csrc/ir/utils.cpp

```diff
@@ -1335,18 +1335,15 @@ void verifyMmaOpForEvaluation(
       "MmaOp::evaluate is not implemented for size: ",
       tv_a->nDims());
 
+  NVF_ERROR(
+      *mma_op->layout() == MmaLayout::TT || *mma_op->layout() == MmaLayout::TN);
```

Comment:
Interesting -- why did you choose to give up NT and NN?

Reply:
`fd.ops.matmul` will plumb down (M,K x K,N) and `fd.ops.linear` will plumb down `M,K x N,K` shapes.

However, now I think `MmaLayout` is not logical shapes but memory layout, which would render this condition incorrect.

Reply:
> now I think MmaLayout is not logical shapes but memory layout

How so? I must have been missing something because https://github.com/NVIDIA/Fuser/blob/9eba1ad122550bf598f4a8d453d003cea989e4fe/tests/cpp/utils.cpp#L332-L347 seems to set up different logical shapes for different so-called `MmaLayout`.

Reply:
Yeah, that's why I also assumed that these govern the logical shapes. But @jacobhinkle pointed out that these are memory layouts in another comment.

@jacobhinkle may be able to offer more clarity.
`mma->layout()` and `getMmaLayout(fusion)`, both return `MmaLayout but it seems they have different definitions at the two places?

Reply:
This comment points to some differences: https://github.com/NVIDIA/Fuser/blob/9eba1ad122550bf598f4a8d453d003cea989e4fe/csrc/scheduler/mma_utils.h#L285

Reply:
There are at least three notions related to layout that need clarifying here I think:
1. The`MmaOp` describes a tensor-core instruction, whose inputs can be transposed in registers or not. There are restrictions on the allowed layouts that those instructions take. During scheduling, this should be set accurately so that `codegen.cpp` generats the proper PTX. This is (I think) always `TN` currently, but it could change in future architectures.
2. The layout returned by `getMmaLayout(fusion)` is the problem layout: i.e. is one of the operands transposed in global memory. There are no restrictions in layouts here. At scheduling, we copy tiles to smem then we load the smem into registers using either `ldmatrix` or `ldmatrix.trans` such that the data is no longer transposed in registers, i.e. we achieve an `MmaOp::layout` equal to `TN`.
3. The test utils are meant to exercise different problem layouts (notion 2). They allocate contiguous arrays with different logical shapes, because we then transpose them to pass them into a mul-sum or matmul op. The matmul op still sees logical shapes of `(M, K) @ (K, N)`.
@zasdfgbnm and @drzejan2 please correct me if any of that is innacurate.

Reply:
> They allocate contiguous arrays with different logical shapes, because we then transpose them to pass them into a mul-sum or matmul op. The matmul op still sees logical shapes of (M, K) @ (K, N).

Thanks! That's what I missed.